### PR TITLE
scipy Rotation and Slerp

### PR DESCRIPTION
**Pull request by chrisflesher**
Squashed PR #15465

### CODE REVIEW

**Comment by chrisflesher** [comment]
@jakevdp Here is the squashed version of the scipy Rotation and Slerp PR.

**Review by jakevdp — APPROVED** [review]
Thanks!

**Comment by jakevdp** [comment]
It looks like there's still a CI error:
```
_________ LaxBackedScipySpatialTransformTests.testRotationConcatenate0 _________
[gw1] linux -- Python 3.8.10 /usr/bin/python3.8
tests/scipy_spatial_test.py:124: in testRotationConcatenate
    self._CheckAgainstNumpy(np_fn, jnp_fn, args_maker, check_dtypes=True, tol=1e-4)
jax/_src/test_util.py:1023: in _CheckAgainstNumpy
    numpy_ans = numpy_reference_op(*args)
tests/scipy_spatial_test.py:123: in <lambda>
    np_fn = lambda q, o: osp_Rotation.concatenate([osp_Rotation.from_quat(q), osp_Rotation.from_quat(o)]).as_quat().astype(dtype)  # HACK
E   AttributeError: type object 'scipy.spatial.transform.rotation.Rotation' has no attribute 'concatenate'
```

**Comment by chrisflesher** [comment]
Ok, removed the `concatenate` function and force pushed an amended commit (so it is still a single commit message).

File: jax/_src/scipy/spatial/transform.py [hunk 1]
```diff
@@ -0,0 +1,393 @@
...
+class Rotation(typing.NamedTuple):
+  """Rotation in 3 dimensions."""
+
+  quat: jax.Array
+
+  # TODO: concatenate is not available in the oldest supported scipy version
+  # @classmethod
+  # def concatenate(cls, rotations: typing.Sequence):
+  #   """Concatenate a sequence of `Rotation` objects."""
+  #   return cls(jnp.concatenate([rotation.quat for rotation in rotations]))
```
Comment (jakevdp) [inline jax/_src/scipy/spatial/transform.py:36]: We should un-comment this to make the API compatible with newer scipy versions.

The problem is not that this is defined; the problem is in the test: we we need to check for the appropriate scipy version before attempting to call `concatenate` on the scipy version of this class.

File: tests/scipy_spatial_test.py [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+  #   shape=[(num_samples, 4)],
+  #   other_shape=[(num_samples, 4)],
+  # )
+  # def testRotationConcatenate(self, shape, other_shape, dtype):
+  #   rng = jtu.rand_default(self.rng())
+  #   args_maker = lambda: (rng(shape, dtype), rng(other_shape, dtype),)
+  #   jnp_fn = lambda q, o: jsp_Rotation.concatenate([jsp_Rotation.from_quat(q), jsp_Rotation.from_quat(o)]).as_quat()
+  #   np_fn = lambda q, o: osp_Rotation.concatenate([osp_Rotation.from_quat(q), osp_Rotation.from_quat(o)]).as_quat().astype(dtype)  # HACK
+  #   self._CheckAgainstNumpy(np_fn, jnp_fn, args_maker, check_dtypes=True, tol=1e-4)
+  #   self._CompileAndCheck(jnp_fn, args_maker, atol=1e-4)
```
Comment (jakevdp) [inline tests/scipy_spatial_test.py:126]: I'd recommend uncommenting this test, but skip it if the scipy version is too old

File: jax/_src/scipy/spatial/transform.py (hunk 1 above)
Comment (chrisflesher) [reply jax/_src/scipy/spatial/transform.py]: Ok, got this working.

File: jax/_src/scipy/spatial/transform.py [hunk 3]
```diff
@@ -0,0 +1,393 @@
...
+  def identity(cls, num: typing.Optional[int] = None, dtype=jnp.float32):
+    """Get identity rotation(s)."""
+    assert num is None
+    quat = jnp.array([0., 0., 0., 1.], dtype=dtype)
+    return cls(quat)
+
+  # @classmethod
+  # def random(cls, random_key: jax.random.PRNGKey, num: typing.Optional[int] = None):
+  #   """Generate uniformly distributed rotations."""
+  #   # Need to implement scipy.stats.special_ortho_group for this to work...
```
[the remark this answers is not in the document]
Comment (chrisflesher) [reply jax/_src/scipy/spatial/transform.py]: Added this.

File: tests/scipy_spatial_test.py (hunk 2 above)
Comment (chrisflesher) [reply tests/scipy_spatial_test.py]: Sounds good, added this.

**Comment by chrisflesher** [comment]
Think all comments are addressed. I had a couple of mypy issues on my local machine that don't seem related to this PR... maybe these won't be a problem for CI? Any advice would be appreciated if this is a real issue.
```
jax/_src/xla_bridge.py:358: error: Cannot find implementation or library stub for module named "importlib_metadata"  [import]
jax/_src/xla_bridge.py:358: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
jax/experimental/jax2tf/call_tf.py:681: error: "list" is not subscriptable, use "typing.List" instead  [misc]
```

**Review by jakevdp — APPROVED** [review]
Thanks!

**Comment by jakevdp** [comment]
It looks like there are a number of mismatched dtype errors when the tests are run in x64 mode. You can reproduce these locally using
```
JAX_ENABLE_X64=1 pytest -n auto tests/scipy_spatial_test.py
```

**Comment by chrisflesher** [comment]
The unit tests are working with 64-bit now (not sure how I missed those before...). The new tests also uncovered a bug in the `_from_matrix` function which should be fixed now.

**Review by jakevdp — APPROVED** [review]
Thanks!

**Comment by jakevdp** [comment]
Last thing, I think: we need a BUILD rule for the new test here: https://github.com/google/jax/blob/0ec9f3c2df808df434e1a620f27cc5d39c06e156/tests/BUILD#L703-L706

**Comment by chrisflesher** [comment]
Ok, modified `jax/BUILD` to include the new test, thanks!

**Review by jakevdp — APPROVED** [review]

File: tests/scipy_spatial_test.py [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+    args_maker = lambda: (rng(shape, dtype),)
+    jnp_fn = lambda q: jsp_Rotation.from_quat(q).single
+    np_fn = lambda q: osp_Rotation.from_quat(q).single
+    self._CheckAgainstNumpy(np_fn, jnp_fn, args_maker, check_dtypes=True, tol=1e-4)
+    self._CompileAndCheck(jnp_fn, args_maker, atol=1e-4)
+
+  @jtu.sample_product(
+    dtype=float_dtypes,
+    shape=[(num_samples, 4)],
+    compute_times=[0., jnp.zeros(1), jnp.zeros(2)],
```
Comment (jakevdp) [inline tests/scipy_spatial_test.py:284]: Sorry, one more issue here. In some test environments, XLA is not yet initialized at module import time, so you cannot create JAX arrays at the main module level. We can fix that by changing these two calls from `jnp.zeros` to `np.zeros`

File: tests/BUILD [hunk 5]
```diff
@@ -727,6 +727,11 @@ jax_test(
     },
 )
 
+jax_test(
+    name = "scipy_spatial_test",
+    srcs = ["scipy_spatial_test.py"],
+)
```
Comment (jakevdp) [inline tests/BUILD:733]: please add here `deps = py_deps("scipy")`

**Review by jakevdp — CHANGES_REQUESTED** [review]
Sorry, a couple more things

**Comment by chrisflesher** [comment]
It looked like some of the tests were failing because the quaternions had an opposite sign. This is actually a correct answer (multiplying all elements of a quaternion by -1 describes the same rotation). Modified the unit tests to use `as_rotvec` instead of `as_quat`, hopefully this will be more robust.

File: tests/BUILD (hunk 5 above)
Comment (chrisflesher) [reply tests/BUILD:733]: Ok.

File: tests/scipy_spatial_test.py (hunk 4 above)
Comment (chrisflesher) [reply tests/scipy_spatial_test.py]: Ok, changed this.

**Comment by chrisflesher** [comment]
Also addressed your last couple of comments.

**Review by jakevdp — APPROVED** [review]
Thanks! We're inching ever closer

**Comment by jakevdp** [comment]
Got it in - thanks so much for the contribution! This was a bit trickier than normal, but I appreciate you sticking with it 😀

**Comment by chrisflesher** [comment]
Cool, thanks for your help! It was fun.

**Pull request merged**